File: agents/shared/python/src/logger.py
"""Structured JSON logger built on structlog.

The REDACT_KEYS list MUST stay in sync with the known-secret env var
names across the platform. Any new secret env var added to any service
goes here too.
"""

from __future__ import annotations

import logging
import sys
from typing import Any

import structlog

REDACT_KEYS: frozenset[str] = frozenset(
    {
        "DATABASE_URL",
        "JWT_SIGNING_KEY",
        "ANTHROPIC_API_KEY",
        "SMS_PROVIDER_TOKEN",
        "EMAIL_DELIVERY_PROVIDER_TOKEN",
        "PERSON_DATA_API_KEY",
        "password",
        "secret",
        "authorization",
        "cookie",
    }
)


def _redact(_logger: Any, _name: str, event_dict: dict[str, Any]) -> dict[str, Any]:
    """Replace values whose key matches REDACT_KEYS with '[redacted]'."""

    def walk(obj: Any) -> Any:
        if isinstance(obj, dict):
            return {
                k: ("[redacted]" if k in REDACT_KEYS else walk(v))
                for k, v in obj.items()
            }
        if isinstance(obj, list):
            return [walk(item) for item in obj]
        return obj

    return walk(event_dict)  # type: ignore[return-value]


def build_logger(*, level: str = "info", service: str | None = None) -> structlog.BoundLogger:
    """Configure structlog with JSON output, ISO timestamps, and redaction."""

    logging.basicConfig(
        format="%(message)s",
        stream=sys.stdout,
        level=getattr(logging, level.upper(), logging.INFO),
    )

    structlog.configure(
        processors=[
            structlog.contextvars.merge_contextvars,
            structlog.processors.add_log_level,
            structlog.processors.TimeStamper(fmt="iso"),
            _redact,
            structlog.processors.JSONRenderer(),
        ],
        wrapper_class=structlog.make_filtering_bound_logger(
            getattr(logging, level.upper(), logging.INFO),
        ),
        cache_logger_on_first_use=True,
    )

    logger = structlog.get_logger()
    if service:
        logger = logger.bind(service=service)
    return logger

File: agents/shared/node/src/logger.js
// Structured JSON logger.
//
// Redact list MUST stay in sync with the known-secret env var names across
// the platform. If a new secret env var is added to any service, add it
// here before committing. The redact paths use pino's glob syntax; each
// entry covers the field appearing anywhere in a log payload.

const pino = require("pino");

const REDACT_PATHS = [
  "*.DATABASE_URL",
  "*.JWT_SIGNING_KEY",
  "*.ANTHROPIC_API_KEY",
  "*.SMS_PROVIDER_TOKEN",
  "*.EMAIL_DELIVERY_PROVIDER_TOKEN",
  "*.PERSON_DATA_API_KEY",
  "*.password",
  "*.secret",
  "*.authorization",
  "*.cookie",
  "req.headers.authorization",
  "req.headers.cookie",
];

/**
 * Build a configured pino logger.
 *
 * @param {object} [opts]
 * @param {string} [opts.level] Override LOG_LEVEL.
 * @param {string} [opts.service] Adds a stable `service` field to every log.
 * @param {string[]} [opts.extraRedactPaths] Service-specific redactions.
 * @returns {import('pino').Logger}
 */
function buildLogger(opts = {}) {
  const { level = process.env.LOG_LEVEL || "info", service, extraRedactPaths = [] } = opts;

  return pino({
    level,
    base: service ? { service } : undefined,
    redact: {
      paths: [...REDACT_PATHS, ...extraRedactPaths],
      censor: "[redacted]",
    },
    timestamp: pino.stdTimeFunctions.isoTime,
  });
}

module.exports = { buildLogger, REDACT_PATHS };
